### PR TITLE
Update scripts/check-v-next-dependencies.js

### PR DESCRIPTION
**Pull request by alcuadrado**
This PR updates the check dependencies script in `v-next` so that it checks `v-next`'s dependency rules.

The main change is that it now enforces that we install the `peerDependencies` of any internal dependency, as pnpm doesn't do it for `workspace:` dependencies.

I also cleaned it up, as updating it was challenging.

### CODE REVIEW

**Comment by alcuadrado** [comment]
@sz-piotr what should we do with this kind of PR, which doesn't trigger `ci` nor `lint`?

**Comment by galargh** [comment]
> @sz-piotr what should we do with this kind of PR, which doesn't trigger `ci` nor `lint`?

@alcuadrado I think we should definitely run v-next CI on PRs modifying `scripts` since we use files from there in the workflow. I added that here, https://github.com/NomicFoundation/hardhat/pull/5675

More generally, I think it might be wise to run `ci` and `lint` on any PR to the `v-next` branch since we require it for the merge queue. I proposed it in https://github.com/NomicFoundation/hardhat/pull/5676

It's worth noting that even without these changes, if we force-merged the PR, the v-next CI workflow would run in the merge queue since `merge_group` doesn't have `paths` filter at all.

Let me know which one you'd like to proceed with.

**Review by galargh — APPROVED** [review]
Looks great :) And what a catch! BTW, as I mentioned on Slack, I also found https://pnpm.io/package_json#dependenciesmetainjected which could be helpful, but all in all, I think that with this in place, we might be even better off 😄

**Comment by alcuadrado** [comment]
> It's worth noting that even without these changes, if we force-merged the PR, the v-next CI workflow would run in the merge queue since merge_group doesn't have paths filter at all.

Ohh, i completely missed this and merged #5676, which looked good to me.

File: scripts/check-v-next-dependencies.js [hunk 1]
```diff
@@ -2,177 +2,273 @@
...
+    .concat(Object.keys(package.devDependencies))
+    .concat(Object.keys(package.peerDependencies));
+
+  for (const dependency of allDependencies) {
+    success &&= validateVersion(
+      package.name,
+      dependency,
+      package.dependencies[dependency],
+      versions
+    );
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply scripts/check-v-next-dependencies.js:90]: Good point! I renamed it

File: scripts/check-v-next-dependencies.js [hunk 2]
```diff
@@ -2,177 +2,273 @@
...
 
-    const peersOk = checkPeerDepedencies(packageJson);
-    const dependencyMap = getDependencyMap(packageJson);
-    dependencyMaps.push(dependencyMap);
+    console.error(`Incompatible versions of dependency: ${dependencyName}`);
 
-    if (peersOk === false) {
-      success = false;
+    for (const [spec, packageNames] of versionSpecs.entries()) {
+      console.log(`  Packages with version ${spec}:`);
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply scripts/check-v-next-dependencies.js:219]: great catch! thanks

File: scripts/check-v-next-dependencies.js [hunk 3]
```diff
@@ -2,177 +2,273 @@
...
-  addDependencies(packageJson.name, packageJson.peerDependencies, dependencies);
+  for (const package of packages.values()) {
+    const allDependencies = Object.entries(package.dependencies)
+      .concat(Object.entries(package.devDependencies))
+      .concat(Object.entries(package.peerDependencies));
 
-  return dependencies;
-}
+    for (const [dependencyName, version] of allDependencies) {
+      if (version.startsWith("workspace:")) {
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply scripts/check-v-next-dependencies.js]: good idea

**Comment by alcuadrado** [comment]
I went ahead and merged this to be able to use the scripts in #5598

**Comment by galargh** [comment]
> > @sz-piotr what should we do with this kind of PR, which doesn't trigger `ci` nor `lint`?
> 
> Why was I pinged?

Oh, hi, Piotr! I think it might have been a mistake, on account of my first name being Piotr, too. Sorry about that

**Pull request merged**